pulp_container/app/serializers.py

```diff
@@ -308,6 +308,24 @@ class ContainerPullThroughRemoteSerializer(RemoteSerializer):
 
     policy = serializers.ChoiceField(choices=[Remote.ON_DEMAND], default=Remote.ON_DEMAND)
 
+    connect_timeout = serializers.FloatField(
+        default=2.0,
+        required=False,
+        help_text=(
+            "aiohttp.ClientTimeout.connect (q.v.) for download-connections. The default is null, "
```

Comment:
the default you're having here is 2

Reply:
while i understand why to decrease the max-retries, I would keep the default timeout for the aiohttp

Reply:
This would mean that a user will need to wait 15 seconds or more (based on the pulpcore's backoff default) to get an error like: "Connection not available". I thought that it makes sense to cap the max value.

Reply:
I am not sure 2 seconds is enough for the overloaded server to create a new connection. Where did you take the 15s and why you think pulp will backoff? It will backoff only if you have set max_retries which you're setting to 0 here.

Reply:
When I turn off the internet connection (I cannot reproduce the timeout thingy), I am seeing this in the logs:
```
pulp_1   | pulp [78399887f2ef439eb8a89b4f18b921c3]: django.request:WARNING: Unauthorized: /v2/
pulp_1   | ('pulp [78399887f2ef439eb8a89b4f18b921c3]: ::ffff:127.0.0.1 - - [29/May/2024:12:01:59 +0000] "GET /v2/ HTTP/1.0" 401 106 "-" "containers/5.29.2 (github.com/containers/image)"',)
pulp_1   | Backing off download_wrapper(...) for 0.2s (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
pulp_1   | pulp [0ad9a5e4b10d4b08a26bc7700984da47]: backoff:INFO: Backing off download_wrapper(...) for 0.2s (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
pulp_1   | Backing off download_wrapper(...) for 0.3s (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
pulp_1   | pulp [0ad9a5e4b10d4b08a26bc7700984da47]: backoff:INFO: Backing off download_wrapper(...) for 0.3s (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
pulp_1   | Backing off download_wrapper(...) for 2.6s (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
pulp_1   | pulp [0ad9a5e4b10d4b08a26bc7700984da47]: backoff:INFO: Backing off download_wrapper(...) for 2.6s (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
pulp_1   | Giving up download_wrapper(...) after 4 tries (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
pulp_1   | pulp [0ad9a5e4b10d4b08a26bc7700984da47]: backoff:ERROR: Giving up download_wrapper(...) after 4 tries (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
pulp_1   | pulp [0ad9a5e4b10d4b08a26bc7700984da47]: django.request:ERROR: Service Unavailable: /v2/docker-cache/library/busybox/manifests/latest
pulp_1   | ('pulp [0ad9a5e4b10d4b08a26bc7700984da47]: ::ffff:127.0.0.1 - admin [29/May/2024:12:03:23 +0000] "GET /v2/docker-cache/library/busybox/manifests/latest HTTP/1.0" 503 92 "-" "containers/5.29.2 (github.com/containers/image)"',)
```

In total, it takes this much time for podman to figure out the remote server is not available:
```
time podman pull localhost:5001/docker-cache/library/busybox --tls-verify=false

Trying to pull localhost:5001/docker-cache/library/busybox:latest...
Error: initializing source docker://localhost:5001/docker-cache/library/busybox:latest: reading manifest latest in localhost:5001/docker-cache/library/busybox: received unexpected HTTP status: 504 Gateway Time-out

real	1m0.116s
```

I believe the default timeout is set to 300 seconds, see the `ClientTimeout` setting argument: https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientSession. Reducing this value as well as `max_retry` might be useful for user.

Reply:
With `connect_timeout` set to 2.0, I get this:
```
pulp_1   | ('pulp [ff2655d508f94c90af9128257e68a8a0]: ::ffff:127.0.0.1 - - [29/May/2024:12:14:29 +0000] "GET /v2/ HTTP/1.0" 401 106 "-" "containers/5.29.2 (github.com/containers/image)"',)
pulp_1   | Backing off download_wrapper(...) for 0.7s (pulpcore.exceptions.base.TimeoutException: Request timed out for https://registry-1.docker.io/v2/library/busybox/manifests/latest. Increasing the total_timeout value on the remote might help.)
pulp_1   | pulp [a40f99af754d45f1b96ef47240b36684]: backoff:INFO: Backing off download_wrapper(...) for 0.7s (pulpcore.exceptions.base.TimeoutException: Request timed out for https://registry-1.docker.io/v2/library/busybox/manifests/latest. Increasing the total_timeout value on the remote might help.)
pulp_1   | Backing off download_wrapper(...) for 1.3s (pulpcore.exceptions.base.TimeoutException: Request timed out for https://registry-1.docker.io/v2/library/busybox/manifests/latest. Increasing the total_timeout value on the remote might help.)
pulp_1   | pulp [a40f99af754d45f1b96ef47240b36684]: backoff:INFO: Backing off download_wrapper(...) for 1.3s (pulpcore.exceptions.base.TimeoutException: Request timed out for https://registry-1.docker.io/v2/library/busybox/manifests/latest. Increasing the total_timeout value on the remote might help.)
pulp_1   | Backing off download_wrapper(...) for 1.3s (pulpcore.exceptions.base.TimeoutException: Request timed out for https://registry-1.docker.io/v2/library/busybox/manifests/latest. Increasing the total_timeout value on the remote might help.)
pulp_1   | pulp [a40f99af754d45f1b96ef47240b36684]: backoff:INFO: Backing off download_wrapper(...) for 1.3s (pulpcore.exceptions.base.TimeoutException: Request timed out for https://registry-1.docker.io/v2/library/busybox/manifests/latest. Increasing the total_timeout value on the remote might help.)
pulp_1   | Giving up download_wrapper(...) after 4 tries (pulpcore.exceptions.base.TimeoutException: Request timed out for https://registry-1.docker.io/v2/library/busybox/manifests/latest. Increasing the total_timeout value on the remote might help.)
pulp_1   | pulp [a40f99af754d45f1b96ef47240b36684]: backoff:ERROR: Giving up download_wrapper(...) after 4 tries (pulpcore.exceptions.base.TimeoutException: Request timed out for https://registry-1.docker.io/v2/library/busybox/manifests/latest. Increasing the total_timeout value on the remote might help.)
pulp_1   | pulp [a40f99af754d45f1b96ef47240b36684]: django.request:ERROR: Service Unavailable: /v2/docker-cache/library/busybox/manifests/latest
pulp_1   | ('pulp [a40f99af754d45f1b96ef47240b36684]: ::ffff:127.0.0.1 - admin [29/May/2024:12:14:40 +0000] "GET /v2/docker-cache/library/busybox/manifests/latest HTTP/1.0" 503 92 "-" "containers/5.29.2 (github.com/containers/image)"',)
```

```
time podman pull localhost:5001/docker-cache/library/busybox --tls-verify=false

Trying to pull localhost:5001/docker-cache/library/busybox:latest...
Error: initializing source docker://localhost:5001/docker-cache/library/busybox:latest: reading manifest latest in localhost:5001/docker-cache/library/busybox: received unexpected HTTP status: 503 Service Unavailable

real	0m11.751s
```

This is what the user sees with the proposed changes. I think it is much more better.

Reply:
It appears like lowering the value of `connect_timeout` has a side effect of giving up the connection sooner, leading to faster error times.

Reply:
Even though `max_retries` is set to 0, I am receiving more than 1 backoff messages. Which is questionable why it is not working in the first place...

Reply:
Because you decreased the value you even get message that says `Increasing the total_timeout value on the remote might help` meaning that it was not enough time to establish the connection( which in this particular case would not work anyway), but you risk with this break even 'normal' cases with internet connections to fail.
I really think you should not  change the timeout connect value or at least leave it 10s ( that seems reasonable enough value but again are you really willing to take the responsibility to set a value which you have not tested how it manifests in connected cases?)